netmap/attributes_test.go

```diff
@@ -0,0 +1,226 @@
+package netmap_test
+
+import (
+	"strconv"
+	"testing"
+
+	"github.com/nspcc-dev/neofs-api-go/v2/netmap"
+	"github.com/nspcc-dev/neofs-api-go/v2/refs"
+	"github.com/stretchr/testify/require"
+)
+
+func subnetAttrKey(val string) string {
+	return "__NEOFS__SUBNET_" + val
+}
+
+func assertSubnetAttrKey(t *testing.T, attr *netmap.Attribute, num uint32) {
+	require.Equal(t, subnetAttrKey(strconv.FormatUint(uint64(num), 10)), attr.GetKey())
+}
+
+func TestWriteSubnetInfo(t *testing.T) {
+	t.Run("entry", func(t *testing.T) {
+		t.Run("zero subnet", func(t *testing.T) {
+			var (
+				node netmap.NodeInfo
+				info netmap.NodeSubnetInfo
+			)
+
+			netmap.WriteSubnetInfo(&node, info)
+
+			// entry to zero subnet does not require an attribute
+			attrs := node.GetAttributes()
+			require.Empty(t, attrs)
+
+			// exit the subnet
+			info.SetEntryFlag(false)
+
+			netmap.WriteSubnetInfo(&node, info)
+
+			// exit from zero subnet should be clearly reflected in attributes
+			attrs = node.GetAttributes()
+			require.Len(t, attrs, 1)
+
+			attr := attrs[0]
+			assertSubnetAttrKey(t, attr, 0)
+			require.Equal(t, "False", attr.GetValue())
+
+			// again enter to zero subnet
+			info.SetEntryFlag(true)
+
+			netmap.WriteSubnetInfo(&node, info)
+
+			// attribute should be removed
+			attrs = node.GetAttributes()
+			require.Empty(t, attrs)
+		})
+
+		t.Run("non-zero subnet", func(t *testing.T) {
+			var (
+				node netmap.NodeInfo
+				info netmap.NodeSubnetInfo
+				id   refs.SubnetID
+			)
+
+			// create non-zero subnet ID
+			const num = 15
+
+			id.SetValue(num)
+
+			// enter to the subnet
+			info.SetID(&id)
+			info.SetEntryFlag(true)
+
+			netmap.WriteSubnetInfo(&node, info)
+
+			// check attribute format
+			attrs := node.GetAttributes()
+			require.Len(t, attrs, 1)
+
+			attr := attrs[0]
+			assertSubnetAttrKey(t, attr, num)
+			require.Equal(t, "True", attr.GetValue())
+
+			// again exit the subnet
+			info.SetEntryFlag(false)
+
+			netmap.WriteSubnetInfo(&node, info)
+
+			// attribute should be removed
+			attrs = node.GetAttributes()
+			require.Empty(t, attrs)
+		})
+	})
+}
+
+func TestSubnets(t *testing.T) {
+	t.Run("empty", func(t *testing.T) {
+		var node netmap.NodeInfo
+
+		called := 0
+
+		err := netmap.IterateSubnets(&node, func(id refs.SubnetID) error {
+			called++
+
+			require.True(t, refs.IsZeroSubnet(&id))
+
+			return nil
+		})
+
+		require.NoError(t, err)
+		require.EqualValues(t, 1, called)
+	})
+
+	t.Run("with correct attribute", func(t *testing.T) {
+		var (
+			node netmap.NodeInfo
+			attr netmap.Attribute
+		)
+
+		attr.SetKey(subnetAttrKey("13"))
+		attr.SetValue("True")
```

Comment:
It would be great to see both `True` and `False` here for both default and non-default subnets.

Reply:
Node attributes must be key-unique according to NeoFS API V2. I've declared undefined behavior of `IterateSubnets` for non-unique attribute keys. The function is not expected to be used outside NeoFS protocol for now.